plugins/catalog-backend-module-github/src/providers/GithubEntityProvider.ts

```diff
@@ -0,0 +1,313 @@
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return schedule.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.options.logger.child({
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(error);
+          }
+        },
+      });
+    };
+  }
+
+  async refresh(logger: Logger) {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    const gitHubConfig = this.integrations.github.byUrl(
+      this.options.orgUrl,
+    )?.config;
+
+    if (!gitHubConfig) {
+      throw new Error(
+        `There is no GitHub integration that matches ${this.options.orgUrl}. Please add a configuration entry for it under integrations.github`,
+      );
+    }
+
+    const { org, repoSearchPath } = parseUrl(this.options.orgUrl);
+
+    const { headers } = await this.githubCredentialsProvider.getCredentials({
+      url: this.options.orgUrl,
+    });
+
+    const client = graphql.defaults({
+      baseUrl: gitHubConfig.apiBaseUrl,
+      headers,
+    });
+
+    const startTimestamp = Date.now();
+
+    logger.info(`Reading GitHub repositories from ${this.options.orgUrl}`);
+
+    if (!this.options.files) {
+      throw new Error(
+        `Missing 'files' array key in 'options' passed to GithubEntityProvider.fromConfig() : ${this.getProviderName()}`,
+      );
+    }
+
+    const expectedEntityFiles = this.options.files;
+
+    const { repositories } = await getOrganizationEntities(
+      client,
+      org,
+      expectedEntityFiles,
+    );
+
+    const matching = repositories.filter(
+      r =>
+        !r.isArchived &&
+        repoSearchPath.test(r.name) &&
+        r.defaultBranchRef?.name,
+    );
+
+    const duration = ((Date.now() - startTimestamp) / 1000).toFixed(1);
+
+    logger.info(
+      `Read ${repositories.length} GitHub repositories (${matching.length} matching the pattern) in ${duration} seconds`,
+    );
+
+    const files = expectedEntityFiles.map(fileName => ({
+      name: fileName,
+      key: camelCase(fileName),
+    }));
+
+    const repoEntities: Array<LocationSpec | Array<LocationSpec>> = [];
+
+    for (const entity of matching) {
+      for (const { name: fileName, key } of files) {
+        const entityObj: any = entity[key as keyof typeof entity];
+        if (entityObj) {
+          const locationEntity = this.createLocationSpec({
+            url: entity.url,
+            branch: entity.defaultBranchRef?.name || '',
+            file: fileName,
+          });
+          repoEntities.push(locationEntity);
+        }
+      }
+    }
+
+    const entityLocations = repoEntities.flat();
+
+    if (entityLocations.length === 0) {
+      logger.info(`No valid entities found by ${this.getProviderName()}.`);
+      return false;
```

Comment:
A minor nit here is that if you ever want to change the return type of this method, it's a breaking change. So if you aren't really very interested in this boolean, you could change the return type to `void` instead. If you really do want to keep track of this boolean, you could change the return type to `{ foundEntities: boolean }` or something like that so you can evolve it and add more fields and stuff (for example the _number of_ things found...) if you like going forward without breaking the contract

Reply:
resolved